### PR TITLE
Implementing support for cuDNN

### PR DESCRIPTION
**Pull request by SergioMartin86**
- Added support for CUDA's cuDNN as backend for NN convolution/normalization/softmax operations
- Fixed a couple of NN hyperparameter initialization bugs in NN that were not breaking but may have affected learning

### CODE REVIEW

File: source/modules/neuralNetwork/layer/layer._cpp [hunk 1]
```diff
@@ -34,15 +34,16 @@ std::vector<float> Layer::generateInitialHyperparameters()
     }
   }
 
-  // If using batch normalization, add layer's batch normalization shift and scale hyperparameters
+  // If using batch normalization, add layer's batch normalization scale and shift hyperparameters
```
Comment (wadaniel) [inline source/modules/neuralNetwork/layer/layer._cpp:37]: these lines resolved the bug?

**Review by wadaniel — APPROVED** [review]

File: source/modules/neuralNetwork/layer/layer._cpp [hunk 2]
```diff
@@ -34,15 +34,16 @@ std::vector<float> Layer::generateInitialHyperparameters()
...
   if (_batchNormalizationEnabled == true)
   {
+    // Adding layer's batch normalization scale hyperparameters
     for (size_t i = 0; i < _nodeCount; i++)
-      hyperparameters.push_back(0.0 + 0.0001 * _nn->_xavierGenerator->getRandomNumber());
```
Comment (webepasc) [inline source/modules/neuralNetwork/layer/layer._cpp:41]: Hoppla now this should have destroyed everthing :O cool how it was able to correct it

File: source/modules/neuralNetwork/neuralNetwork._cpp [hunk 3]
```diff
@@ -280,181 +596,294 @@ std::vector<float> NeuralNetwork::getHyperparameters()
...
 
-#ifdef _KORALI_USE_ONEDNN
-
   size_t layerCount = _layers.size();
   size_t currPos = 0;
   for (size_t i = 0; i < layerCount; i++)
   {
-    const memory::dim IC = _layers[i - 1]->_nodeCount;
-    const memory::dim OC = _layers[i]->_nodeCount;
+    size_t IC = _layers[i - 1]->_nodeCount;
```
Comment (webepasc) [inline source/modules/neuralNetwork/neuralNetwork._cpp:603]: This should go into the i>0 branch, no?

File: source/modules/neuralNetwork/neuralNetwork._cpp [hunk 4]
```diff
@@ -206,37 +363,196 @@ void NeuralNetwork::backward(const std::vector<float> &outputDifferential)
...
+
+#ifdef _KORALI_USE_ONEDNN
+
       _layers[i]->_backwardDataPrimitive.execute(_stream, _layers[i]->_backwardDataArgs);
+
+#endif
+
+#ifdef _KORALI_USE_CUDNN
+
+      float alpha = 1.0f;
```
Comment (webepasc) [inline source/modules/neuralNetwork/neuralNetwork._cpp:486]: shouldn't this be the trainable parameters?

**Review by webepasc — CHANGES_REQUESTED** [review]
Can you check my comments?

File: source/modules/neuralNetwork/layer/layer._cpp (hunk 1 above)
Comment (SergioMartin86) [reply source/modules/neuralNetwork/layer/layer._cpp:37]: lol very funny

File: source/modules/neuralNetwork/neuralNetwork._cpp [hunk 5]
```diff
@@ -96,26 +107,25 @@ std::vector<float> NeuralNetwork::generateInitialHyperparameters()
   for (size_t i = 0; i < _layers.size(); i++)
   {
     auto layerParameters = _layers[i]->generateInitialHyperparameters();
-    initialHyperparameters.insert(initialHyperparameters.begin(), layerParameters.begin(), layerParameters.end());
+    initialHyperparameters.insert(initialHyperparameters.end(), layerParameters.begin(), layerParameters.end());
```
[the remark this answers is not in the document]
Comment (SergioMartin86) [reply source/modules/neuralNetwork/neuralNetwork._cpp:110]: Yes, more or less that. But see that, since layers had different sizes, this would cause some hyperparmeters for the mean/variance to go into shift/scale (and viceversa), which is disastrous (a 0.0 normalization mean is ok, a 0.0 variance is NOT ok).

File: source/modules/neuralNetwork/layer/layer._cpp (hunk 2 above)
Comment (SergioMartin86) [reply source/modules/neuralNetwork/layer/layer._cpp:41]: Yeah, this was also part of the problem

File: source/modules/neuralNetwork/neuralNetwork._cpp (hunk 3 above)
Comment (SergioMartin86) [reply source/modules/neuralNetwork/neuralNetwork._cpp:603]: yes, but I think you or Daniel had solved this in the branch you are working on now?
